zebra-chain/src/orchard/keys.rs

```diff
@@ -0,0 +1,1009 @@
+    network: Network,
+    spend_validating_key: SpendValidatingKey,
+    nullifier_deriving_key: NullifierDerivingKey,
+    ivk_commit_randomness: IvkCommitRandomness,
+}
+
+// TODO: impl a From that accepts a Network?
+
+impl fmt::Debug for FullViewingKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("FullViewingKey")
+            .field("network", &self.network)
+            .field("spend_validating_key", &self.spend_validating_key)
+            .field("nullifier_deriving_key", &self.nullifier_deriving_key)
+            .field("ivk_commit_randomness", &self.ivk_commit_randomness)
+            .finish()
+    }
+}
+
+impl fmt::Display for FullViewingKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut bytes = io::Cursor::new(Vec::new());
+
+        let _ = bytes.write_all(&<[u8; 32]>::from(self.spend_validating_key));
+        let _ = bytes.write_all(&<[u8; 32]>::from(self.nullifier_deriving_key));
+        let _ = bytes.write_all(&<[u8; 32]>::from(self.ivk_commit_randomness));
+
+        let hrp = match self.network {
+            Network::Mainnet => fvk_hrp::MAINNET,
+            Network::Testnet => fvk_hrp::TESTNET,
+        };
+
+        bech32::encode_to_fmt(f, hrp, bytes.get_ref().to_base32(), Variant::Bech32).unwrap()
+    }
+}
+
+impl From<SpendingKey> for FullViewingKey {
+    fn from(sk: SpendingKey) -> FullViewingKey {
+        let spend_authorizing_key = SpendAuthorizingKey::from(sk);
+
+        Self {
+            // TODO: handle setting the Network better.
+            network: Network::default(),
+            spend_validating_key: SpendValidatingKey::from(spend_authorizing_key),
+            nullifier_deriving_key: NullifierDerivingKey::from(sk),
+            ivk_commit_randomness: IvkCommitRandomness::from(sk),
+        }
+    }
+}
+
+impl FromStr for FullViewingKey {
+    type Err = SerializationError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match bech32::decode(s) {
+            Ok((hrp, bytes, Variant::Bech32)) => {
+                let mut decoded_bytes = io::Cursor::new(Vec::<u8>::from_base32(&bytes).unwrap());
+
+                let ak_bytes = decoded_bytes.read_32_bytes()?;
+                let nk_bytes = decoded_bytes.read_32_bytes()?;
+                let rivk_bytes = decoded_bytes.read_32_bytes()?;
+
+                Ok(FullViewingKey {
+                    network: match hrp.as_str() {
+                        fvk_hrp::MAINNET => Network::Mainnet,
+                        _ => Network::Testnet,
+                    },
+                    spend_validating_key: SpendValidatingKey::from(ak_bytes),
+                    nullifier_deriving_key: NullifierDerivingKey::from(nk_bytes),
+                    ivk_commit_randomness: IvkCommitRandomness::try_from(rivk_bytes).unwrap(),
+                })
+            }
+            _ => Err(SerializationError::Parse("bech32 decoding error")),
+        }
+    }
+}
```

Comment:
Orchard Unified Addresses

Reply:
Keeping as-is for now and will revise as Unified Addresses is finalized: https://github.com/zcash/zips/pull/485